### PR TITLE
create specification for self.browser

### PR DESCRIPTION
**Pull request by patrickkettner**
Create specification for window.browser/self.browser.

preview available [here](https://patrickkettner.github.io/window.browser/)

### CODE REVIEW

**Comment by patrickkettner** [comment]
@Rob--W @xeenon would yall be able to review when time permits?

File: specification/window.browser.html
Comment (xeenon) [inline specification/window.browser.html]: `short` is the wrong type for this. `object` would be the best type to use here until we have a `Namespace` object type spec.

File: specification/window.browser.html
Comment (xeenon) [inline specification/window.browser.html]: Not sure why `orientation` is referenced here.

File: specification/window.browser.html
Comment (xeenon) [inline specification/window.browser.html]: `window` should should likely reference `https://html.spec.whatwg.org/#window` instead of the Fetch spec.

**Review by xeenon — CHANGES_REQUESTED** [review]
Looks good! Some references need corrected to point in the right places.

Also, do you have the original bike shed file? We have Bikeshed transformation setup in our repo, so you should be able to just post that file, like [`index.bs`](https://github.com/w3c/webextensions/blob/main/index.bs) which is more human friendly. That gets published [here](https://w3c.github.io/webextensions/specification).

**Comment by patrickkettner** [comment]
thanks @xeenon!
I added the bikeshed file. The current build system is using [pr-preview](https://github.com/tobie/pr-preview), and [configured](https://github.com/w3c/webextensions/blob/ff053a84ac68eb6e6d8f8df3704eeb7f6631f23c/.pr-preview.json#L2C1-L2C1) to build the [existing spec](https://w3c.github.io/webextensions/specification/). pr-preview has a [long running request](https://github.com/tobie/pr-preview/issues/18) to support multiple files, but the existing PR has been [untouched for years](https://github.com/tobie/pr-preview/pull/95). @dotproto and I discussed replacing the build system with something that supports multiple files built directly on bikeshed, rather than using pr-preview. I plan on digging into it soon. In the meantime, given devlin's desire to give a few months notice to developer before any change landing, I didn't want to block on that.

File: specification/window.browser.html
Comment (patrickkettner) [reply specification/window.browser.html]: removed. It was vestigial code from the compat spec, thanks for pointing it out

File: specification/window.browser.html
Comment (patrickkettner) [reply specification/window.browser.html]: updated

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]: I suppose that the personal Github repositories in the metadata is only temporarily, for preview purposes, right? That is, before merging, this URL and other references will point to the WECG repo.

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]:
```suggestion
   attribute object browser;
```

Suggestion to remove `readonly`, for the following reasons:

- These properties are currently not read-only, at least not in Chrome, nor in Firefox.
- Making it writable reduces the webcompat risk. Using `readonly` breaks websites that assign to `browser`.
- Making it writable also makes it easier to replace the `browser` namespace in case one wants to replace it altogether to support unit tests.

A counterpoint, in favor of adding `readonly`, is that it enables web pages to rely on the presence of an unmodified `browser`, unaffected by other browser extensions or third-party libraries, for example. I am not attributing much value to this feature, because one cannot expect any protections once third-party code has executed in the context.

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]: How about extension service workers?

Do we want to reserve `browser` in regular workers as well?

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs:30]: Question: Is this sufficient, or do we need a very brief introduction to what "WebExtensions" is, to people unfamiliar with WebExtensions?

Originally, it was just "Chrome Extensions". When Mozilla adopted the same model, the name "WebExtensions" was coined, described at https://wiki.mozilla.org/Add-ons/Terminology. Although that page asserts that "WebExtensions" ought to refer to the technology, and not the concrete extensions built with it, in practice the term is being used to refer to both.

Here is a small suggestion to add some context, and to also call out why there is a need for a "browser" namespace at all.

```suggestion
  Browser extensions are programs to extend the capabilities of web browsers.
  Although not standardized at first, web browsers have independently implemented a common interoperable set of extension APIs, called WebExtensions.
  
  Modern WebExtensions are an evolution of a system originally introduced in
```

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: yes, that is correct

File: specification/window.browser.bs
Comment (Rob--W) [reply specification/window.browser.bs]: Although my suggested edit has been approved, I'd like to temporarily re-open this thread to allow other vendors to chime in. In particular, I'd like to hear @xeenon 's perspective on Safari's behavior. I see that Safari's interface of the `browser` namespace content itself is defined at https://github.com/WebKit/WebKit/blob/ea5e30e46342b0e755281a1ffb2f23c37a2b9bcd/Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl , with widespread use of "readonly", but cannot tell from the WebKit source code whether the `browser` global is supposed to be `readonly` or writable.

Comment (xeenon) [reply specification/window.browser.bs]: window.browser is not readonly in WebKit / Safari.

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: added

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: it is also not readonly in Chrome. I would be surprised if there was interest in changing it.

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: updated to the presumed future location

File: specification/window.browser.bs
Comment (oliverdunk) [inline specification/window.browser.bs]:
```suggestion
  That <code>browser</code> is now considered the canonically correct namespace
```

File: specification/window.browser.bs
Comment (oliverdunk) [inline specification/window.browser.bs]: This can happen from a content script or just a website itself so I think avoiding the use of an extension-specific term makes this clearer to outside readers.

```suggestion
  external messaging from websites to background processes) necessitate
```

File: specification/window.browser.bs
Comment (oliverdunk) [inline specification/window.browser.bs:58]: Should we make it clear that this is sometimes undefined, such as if a site is not included in the `externally_connectable` array in any extension's manifest? I'm guessing the right syntax based on [this](https://hg.mozilla.org/integration/autoland/file/d65760500d063d80f6cc3c1fa8c61bfe1ac81412/dom/bindings/test/TestExampleGen.webidl#l582) but it seems to be a new feature that doesn't appear in many real specs so not sure if there's a better way. cc/ @Rob--W @xeenon 

```suggestion
   attribute (undefined or object) browser;
```

File: specification/window.browser.bs
Comment (oliverdunk) [inline specification/window.browser.bs]:
```suggestion
    It MUST be used exclusively for WebExtension purposes, however the contents of each instance of {{browser}} is UA defined.
```

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Thanks so much for putting this together, I'm excited! Left a couple of thoughts but this generally looks good.

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs:58]: it feels weird to set the only attribute being specified to undefined in the spec. Certainly can add it, but I am not sure I see the value if it is user agent defined. Would love to hear Rob's and Timothy's thoughts

**Comment by Jack-Works** [comment]
are you considering renaming this PR (or part of the non-normative sections) to `self.browser`? that covers service workers.

File: specification/window.browser.bs
Comment (dotproto) [reply specification/window.browser.bs:58]: Since WebIDL treats `object` and `undefined` as distinct types, it seems like this would more accurately represent the possible values a user agent may assign the global we're reserving. 

Would it be more appropriate (or conventional) to use a union type as @oliverdunk suggested, a nullable type, or possibly even to declare a (partial) namespace?

```webidl
partial interface Window {
 attribute object? browser;
};
```

```webidl
[Exposed=(Window,Worker)]
partial namespace browser {}
```

I tried looking around at other specs to see if/when/how `namespace` was used and was surprised at how rarely it appeared. The only notable place I saw it used was the [console spec](https://console.spec.whatwg.org/#console-namespace).

Comment (oliverdunk) [reply specification/window.browser.bs:58]: As a note, my (uneducated) reading of the spec was that nullable types were equivalent to `null | something` so I'm not sure they'd be appropriate here. Still very unsure what the right path is though.

**Review by xeenon — APPROVED** [review]

**Review by dotproto — APPROVED** [review]
LGTM

File: specification/window.browser.bs
Comment (oliverdunk) [reply specification/window.browser.bs:58]: We've agreed to revisit this after landing.

Comment (patrickkettner) [reply specification/window.browser.bs:58]: @oliverdunk can you LGTM?

**Review by oliverdunk — APPROVED** [review]
LGTM 👍

**Comment by patrickkettner** [comment]
@Rob--W any thoughts?

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]: This renders as is, with `{#introduction}`. What was the intended result?

File: specification/window.browser.html
Comment (Rob--W) [inline specification/window.browser.html]: I looked at the preview, which features this "exclusivly" typo, which is not present in the .bs file. Can you re-generate the html file to make sure that it matches the .bs file?

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]:
```suggestion
```

AFAIK there are no browser implementations that expose `chrome`/`browser` to the global scope of `Worker`s, not even those created from an extension origin. Therefore I suggest to remove the mention of `DedicatedWorkerGlobalScope`.

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]:
```suggestion
    When {{browser}} is defined on {{window}}, it SHOULD also be exposed on {{ServiceWorkerGlobalScope}}
    of origins associated with WebExtensions.
```

Suggested edit (dropped the dedicated worker as explained in another comment), and emphasized that the origin of the scope must be a WebExtension origin. Also changed "implemented" to "exposed", because "implemented" seems more vague.

Currently we don't have a separate definition for extension workers, but if we ever do (as suggested in #11), then that would be the type to extend instead.

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]: Question: The mention of Service Workers here may raise questions, because Service workers are defined for http(s) URLs only: https://w3c.github.io/ServiceWorker/#start-register-algorithm
(in practice mainly https, http is only supported for edge cases such as localhost).

Chrome (and Firefox's experimental implementation) does not permit extension code to register arbitrary service workers. Instead, the extension framework takes care of registering workers despite the http(s) restrictions.

Should we clarify in a note that extensions cannot register Service workers as usual (i.e. the https restriction), but that browsers may spawn a Service worker associated with a WebExtension origin?

Should we drop the mention of workers entirely from this "window.browser" document, and define the worker behavior in an entirely separate spec/PR?

File: specification/window.browser.bs
Comment (Rob--W) [inline specification/window.browser.bs]: Is it that relevant to mention "version 3" after "Chrome"? I suggest to drop it so that there is no confusion with "Manifest version 3".

(also rename Chrome to Chromium because the rest of the paragraph refers back to it as Chromium.)

```suggestion
  Chromium. In that system, any API that was added was put under the
```

**Comment by Rob--W** [comment]
I know that we wanted us to spec the existence of `browser` because Chrome considers a spec as a prerequisite to exposing this. The specific behavior of "browser" is left up to the UA, so my question here is not a blocking part of the review. Still, I'm curious about the ideas behind it.

Chrome (and Safari) both have the concept of "isolated world" for content scripts. An isolated world shares the same view of the DOM (DOM instances from native code are shared), with unique JavaScript wrappers (isolated per world). If `window.browser` were to be defined in the idl as part of `window`, what does that say about the value read by the main world vs isolated world? Is it the same (native) object, but with different wrappers that somehow filters the available extension API methods? Or is the `browser` namespace a different object (and therefore implicitly also a separate wrapper if any)?

Here is some background on isolated world and a description of bindings: https://chromium.googlesource.com/chromium/src.git/+/123.0.6266.0/third_party/blink/renderer/bindings/core/v8/V8BindingDesign.md

In Firefox, externally_connectable has not been implemented yet ([bug 1319168](https://bugzilla.mozilla.org/show_bug.cgi?id=1319168)), which means that there is no `browser` namespace in the main world, only in that of content scripts. The current implementation of the sandbox in Firefox is that `globalThis` is a sandbox inheriting from `window`. `globalThis.browser` is the extension API namespace, `window.browser` is currently undefined. I imagine that if we were to implement `window.browser`, that the main world version would be generated through WebIDL, and that the content script continues to have its own implementation-defined extension API namespace separate from `window.browser`.

**Comment by xeenon** [comment]
The Safari and WebKit implementations of web extensions currently dynamically add the `browser` object as needed to the DOM global object, per world. So it is always a different wrapper _and_ a different internal object.

It needed to be different since the things contained in the `browser` namespace for the main world in web pages is drastically reduced and different from the extension content script world or main world in extensions pages.

We will likely continue to dynamically add this property to the global scope, instead of adding it in our DOM window IDL.

I don't think that fundamentally blocks anything here. Browser implementations can expose the APIs however their implementation requires, as long as it matches how the IDL in the spec defines it.

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: ah, thanks for pointing that out. That is the syntax for declaring a custom ID in markdown. But bikeshed doens't support it. The correct thing to do would be to define an ID on the section on the line above. But since I added it to be able to link to it later in the doc, and that link was factored out, I will just remove this.

File: specification/window.browser.bs
Comment (patrickkettner) [reply specification/window.browser.bs]: Thanks for the feedback @Rob--W!

> because Service workers are defined for http(s) URLs only: https://w3c.github.io/ServiceWorker/#start-register-algorithm (in practice mainly https, http is only supported for edge cases such as localhost).

I am not sure I follow. The spec in this PR says that if `window.browser` is defined, then `self.browser` should be defined on any `ServiceWorkerGlobalScope` whose `origin` is a webextension. While the service worker spec is only http(s), the thing that is registered as a background script in Chrome for manifest v3 _is_ 
 `ServiceWorkerGlobalScope`. 

> Chrome (and Firefox's experimental implementation) does not permit extension code to register arbitrary service workers. Instead, the extension framework takes care of registering workers despite the http(s) restrictions.

If they are not permitted, then I don't see the problem? ServiceWorkerGlobalScope is what is created when you have an extension in Chrome/Firefox's-experimental-implementation. Can you explain what you are wanting changed?

> Should we clarify in a note that extensions cannot register Service workers as usual (i.e. the https restriction)

While I see the use for that in a broader web extensions specification, as there is no other portion of this document is about registering a service worker, or when it should be registered on `window`. That seems outside of this document's scope.

> Should we drop the mention of workers entirely from this "window.browser" document, and define the worker behavior in an entirely separate spec/PR?

Given the current level of web extension specification, having two essentially identical documents to accomplish the same thing ("`globalThis.browser` should only be used for web extensions") feels unnecessary.

File: specification/window.browser.html
Comment (patrickkettner) [reply specification/window.browser.html]: done

File: specification/window.browser.bs
Comment (Rob--W) [reply specification/window.browser.bs]: > Thanks for the feedback @Rob--W!
> 
> > because Service workers are defined for http(s) URLs only: https://w3c.github.io/ServiceWorker/#start-register-algorithm (in practice mainly https, http is only supported for edge cases such as localhost).
> 
> I am not sure I follow. The spec in this PR says that if `window.browser` is defined, then `self.browser` should be defined on any `ServiceWorkerGlobalScope` whose `origin` is a webextension. While the service worker spec is only http(s), the thing that is registered as a background script in Chrome for manifest v3 _is_ 
>  `ServiceWorkerGlobalScope`. 

In the realm of specifications, Service Workers (and thus ServiceWorkerGlobalScope instances) only exist on https schemes (and localhost). Here we specify that ServiceWorkerGlobalScope should have a browser object, for WebExtension origins only. I initially found this inconsistent, because the SW spec doesn't leave room for SW on other schemes, while our spec here implies the existence of an extension origin.

I have now reconciled this inconsistency by observing that "extension origin" could theoretically also involve one of the accepted schemes (even though that doesn't happen in practice - as an example from a far past, Chrome had the concept of [Hosted apps](https://web.archive.org/web/20191221070723/https://developer.chrome.com/webstore/hosted_apps) where regular websites received extra privileges if a crx file defining the "hosted app" was installed).
We (browser vendors) understand the intent and practical meaning here, and I am fine with not mentioning anything more about SW for now.

**Review by Rob--W — APPROVED** [review]
LGTM! I can't easily verify the preview, but the source looks good!

**Pull request merged**
